designdocs/state_storage.md

```diff
@@ -0,0 +1,116 @@
+# On Chain State Storage
+
+## Context and Scope
+The proposed feature consists of changes that is going to be one (or more) pallet(s) in runtime of a
+Substrate based blockchain, and it will be used in all environments including production.
+
+## Problem Statement
+Every party using any medium of communication will need to deal with the storage of some kind of
+stateful data. Some of these stateful data might be consensus critical and in a decentralized world
+one option might be to store this type of data on a blockchain as long as it is required.
+The scope for how long some of these stateful data should exist is dependent on the
+type of data. For example, some of them might exist as long as an active channel exists between two
+communicating parties and for some others it might be directly dependent on the existence of a
+particular party.
+
+### DSNP Usecase
+**Frequency** is the first implementation of [DSNP](https://dsnp.org) and most of the state
+transitions in **DSNP** can be modeled via `Announcements`. For some of these state transitions we
+only care about the latest state, and currently the only way to achieve this via announcements is to
+have some kind of third party indexer to track the latest state for these types.
+If there was a way to be able to mutate the state of a storage, then we could update it to its
+latest version and potentially remove the need for an indexer.
+
+## Goals
+1. Allowing storage of stateful data with flexible schemas on chain.
+2. Data stored for any `MSA` should not have any effects like changing read/write access times or
+storage costs for any data stored for another `MSA`.
+3. Allowing high write throughput.
+4. Allowing high read throughput.
+5. Avoiding data races between consecutive updates of the same item.
+
+## Non-goals
+- Defining an economic incentive to restrict infinite growth of on-chain database.
+
+## Glossary
+* `DSNP`: Decentralized Social Networking Protocol
+* `MSA`: Message Source Account. A registered identifier with the MSA pallet.
+
+### Modes of Storage
+- **Itemized**: In this mode we are storing each entity of that schemaId as an item in a blob with a
+maximum size of `MAX_ITEMIZED_BLOB_SIZE`.
+This is recommended for entities witch have relatively small size compared to
+`MAX_ITEMIZED_BLOB_SIZE` and there are more than a few of them.
+- **Paginated**: In this mode we are storing each entity individually and inside a separate page.
+This mode is recommended for entities that are relatively large and there are a few of them.
+
+![image](https://user-images.githubusercontent.com/9152501/213291600-98229ee4-6358-4e0e-abe2-d6da9abe179e.png)
+
+The rationale behind separating these modes is based on two reasons:
+- _Performance_: **Itemized** mode always needs 1 on-chain DB_READ and 1 DB_WRITE but **Paginated**
+mode might be done with only 1 DB_WRITE.
+- _Composability_: **Itemized** mode allows better composability due to on-chain read requirement
+and smaller size of each.
```

Comment:
I think in terms of graph that is correct but since this is a generic storage there might be usecases that you might not need to read all pages.
One other thing to remember is that since reading is an off-chain process it does not have any impact on performance of the node.